src/cargo/core/compiler/job.rs

```diff
@@ -7,6 +8,8 @@ use crate::util::CargoResult;
 pub struct Job {
     work: Work,
     fresh: Freshness,
+
+    dirty_reason: Option<DirtyReason>,
```

Comment:
That was actually my first idea(and I asked on zulip too), but there are some places where `Job::new_dirty()` is called with `None` as the `Option<DirtyReason>`, so it would need to be `Dirty(Option<DirtyReason>)`, and it doesn't really make that much of a difference then.

Reply:
For `fn note_working_on`, there might be a discrepancy that `Freshness` is `Fresh` while we still accidentally pass in `Some(DirtyReason)`.

Not a blocker though. Just realize we may encode the logic more precisely into the type system.

Reply:
Yeah, that's a good point. I'm going to update that then.

Reply:
Done in 00b81c2e7295c619356293d9abfcb01dcd51205c